src/peft/tuners/lora/tp_layer.py

```diff
@@ -0,0 +1,155 @@
+            parallel_linear: Union[tensor_parallel.ColumnParallelLinear, tensor_parallel.RowParallelLinear],
+            r: int = 0,
+            lora_alpha: int = 1,
+            lora_dropout: float = 0.0,
+            fan_in_fan_out: bool = False,
+            # Set this to True if the layer to replace stores weight like (fan_in, fan_out)
+            **kwargs,
+    ):
+        init_lora_weights = kwargs.pop("init_lora_weights", True)
+
+        self.parallel_linear_class = type(parallel_linear)
+        config = MegatronConfig.config
+        parallel_linear_kwargs = {'config': config}
+        if hasattr(config, 'init_method'):
+            parallel_linear_kwargs['init_method'] = config.init_method
+        if isinstance(parallel_linear, tensor_parallel.RowParallelLinear):
+            parallel_linear_kwargs['input_is_parallel'] = parallel_linear.input_is_parallel
+        else:
+            parallel_linear_kwargs['gather_output'] = parallel_linear.gather_output
+        type(parallel_linear).__init__(self, input_size=parallel_linear.input_size,
+                                       output_size=parallel_linear.output_size, bias=parallel_linear.bias is not None,
+                                       skip_bias_add=parallel_linear.skip_bias_add,
+                                       **parallel_linear_kwargs)
+
+        LoraParalleLayer.__init__(self, in_features=parallel_linear.input_size,
+                                  out_features=parallel_linear.output_size,
+                                  is_paralle_a=isinstance(parallel_linear, tensor_parallel.RowParallelLinear))
+
+        self.fan_in_fan_out = fan_in_fan_out
+
+        self.update_layer(adapter_name, r, lora_alpha, lora_dropout, init_lora_weights, **parallel_linear_kwargs)
+        self.is_target_conv_1d_layer = False
+        self.set_adapter(adapter_name)
+
+    def forward(self, x: torch.Tensor):
+        previous_dtype = x.dtype
+        # If weight is used for matrix multiplication here, the final aggregation operation of the original
+        # parallel_linear layer will be missing, so we need to directly call its forward function to obtain the
+        # output of the original parallel_linear layer.
+        if self.disable_adapters:
+            if self.merged:
+                self.unmerge()
+            result, bias = self.parallel_linear_class.forward(self, x)
+        elif self.merged:
+            result, bias = self.parallel_linear_class.forward(self, x)
+        else:
+            result, bias = self.parallel_linear_class.forward(self, x)
+            for active_adapter in self.active_adapters:
+                if active_adapter not in self.lora_A.keys():
+                    continue
+                lora_A = self.lora_A[active_adapter]
+                lora_B = self.lora_B[active_adapter]
+                dropout = self.lora_dropout[active_adapter]
+                scaling = self.scaling[active_adapter]
+                x = x.to(lora_A.weight.dtype)
+
```

Comment:
not see amp_02 with apex for layer norm
linear_in can use ColumnParallelLinear 
linear_out can use  RowParallelLinear if gather_out

Reply:
> not see amp_02 with apex for layer norm linear_in can use ColumnParallelLinear linear_out can use RowParallelLinear if gather_out

As the first reason, lora matrix is small, so if we use linear, the more copy in gpu has little influence, but if we use tensor_parallel, we will add one more scatter op, the communication overhead is big, so the efficiency may not necessarily be improved.